File: web/src/lib/schema.ts
import {
  SCHEMA_VERSION,
  type ActionSequence,
  type ForgeAction,
  SAFETY_LIMITS,
} from '../types/actions';

export class SchemaValidationError extends Error {
  constructor(message: string) {
    super(message);
    this.name = 'SchemaValidationError';
  }
}

function clamp(value: number, min: number, max: number): number {
  return Math.min(max, Math.max(min, value));
}

export function normalizeAction(action: ForgeAction): ForgeAction {
  switch (action.type) {
    case 'move_mouse':
      return {
        ...action,
        duration: clamp(
          action.duration ?? 0.5,
          SAFETY_LIMITS.minMouseDuration,
          SAFETY_LIMITS.maxMouseDuration,
        ),
      };
    case 'type_text':
      return {
        ...action,
        interval: clamp(
          action.interval ?? 0.05,
          SAFETY_LIMITS.minTypeInterval,
          SAFETY_LIMITS.maxTypeInterval,
        ),
      };
    case 'wait':
      return {
        ...action,
        seconds: clamp(action.seconds, 0.1, SAFETY_LIMITS.maxWaitSeconds),
      };
    case 'click':
      return { ...action, button: action.button ?? 'left' };
    default:
      return action;
  }
}

export function normalizeSequence(sequence: ActionSequence): ActionSequence {
  return {
    version: SCHEMA_VERSION,
    name: sequence.name,
    description: sequence.description,
    actions: sequence.actions.map(normalizeAction),
  };
}

export function validateAction(action: unknown): ForgeAction {
  if (!action || typeof action !== 'object') {
    throw new SchemaValidationError('Action must be an object');
  }

  const a = action as Record<string, unknown>;
  const type = a.type;

  if (typeof type !== 'string') {
    throw new SchemaValidationError('Action type is required');
  }

  switch (type) {
    case 'move_mouse':
      if (typeof a.x !== 'number' || typeof a.y !== 'number') {
        throw new SchemaValidationError('move_mouse requires numeric x and y');
      }
      return normalizeAction({
        type: 'move_mouse',
        x: a.x,
        y: a.y,
        duration: typeof a.duration === 'number' ? a.duration : 0.5,
        label: typeof a.label === 'string' ? a.label : undefined,
      });
    case 'click':
      return normalizeAction({
        type: 'click',
        x: typeof a.x === 'number' ? a.x : undefined,
        y: typeof a.y === 'number' ? a.y : undefined,
        button: a.button === 'right' || a.button === 'middle' ? a.button : 'left',
        label: typeof a.label === 'string' ? a.label : undefined,
      });
    case 'double_click':
      return {
        type: 'double_click',
        x: typeof a.x === 'number' ? a.x : undefined,
        y: typeof a.y === 'number' ? a.y : undefined,
        label: typeof a.label === 'string' ? a.label : undefined,
      };
    case 'right_click':
      return {
        type: 'right_click',
        x: typeof a.x === 'number' ? a.x : undefined,
        y: typeof a.y === 'number' ? a.y : undefined,
        label: typeof a.label === 'string' ? a.label : undefined,
      };
    case 'type_text':
      if (typeof a.text !== 'string') {
        throw new SchemaValidationError('type_text requires text');
      }
      return normalizeAction({
        type: 'type_text',
        text: a.text,
        interval: typeof a.interval === 'number' ? a.interval : 0.05,
        label: typeof a.label === 'string' ? a.label : undefined,
      });
    case 'press_key':
      if (typeof a.key !== 'string') {
        throw new SchemaValidationError('press_key requires key');
      }
      return {
        type: 'press_key',
        key: a.key,
        label: typeof a.label === 'string' ? a.label : undefined,
      };
    case 'hotkey':
      if (!Array.isArray(a.keys) || a.keys.length < 2) {
        throw new SchemaValidationError('hotkey requires at least 2 keys');
      }
      return {
        type: 'hotkey',
        keys: a.keys as string[],
        label: typeof a.label === 'string' ? a.label : undefined,
      };
    case 'wait':
      if (typeof a.seconds !== 'number') {
        throw new SchemaValidationError('wait requires seconds');
      }
      return normalizeAction({
        type: 'wait',
        seconds: a.seconds,
        label: typeof a.label === 'string' ? a.label : undefined,
      });
    case 'open_application':
      if (typeof a.target !== 'string') {
        throw new SchemaValidationError('open_application requires target');
      }
      return {
        type: 'open_application',
        target: a.target,
        label: typeof a.label === 'string' ? a.label : undefined,
      };
    case 'scroll':
      if (typeof a.amount !== 'number') {
        throw new SchemaValidationError('scroll requires amount');
      }
      return {
        type: 'scroll',
        amount: a.amount,
        x: typeof a.x === 'number' ? a.x : undefined,
        y: typeof a.y === 'number' ? a.y : undefined,
        label: typeof a.label === 'string' ? a.label : undefined,
      };
    default:
      throw new SchemaValidationError(`Unknown action type: ${type}`);
  }
}

export function validateSequence(input: unknown): ActionSequence {
  if (!input || typeof input !== 'object') {
    throw new SchemaValidationError('Sequence must be an object');
  }

  const seq = input as Record<string, unknown>;
  const actions = seq.actions;

  if (!Array.isArray(actions) || actions.length === 0) {
    throw new SchemaValidationError('Sequence must have at least one action');
  }

  return normalizeSequence({
    version: SCHEMA_VERSION,
    name: typeof seq.name === 'string' ? seq.name : undefined,
    description: typeof seq.description === 'string' ? seq.description : undefined,
    actions: actions.map(validateAction),
  });
}

export function serializeSequence(sequence: ActionSequence): string {
  return JSON.stringify(normalizeSequence(sequence), null, 2);
}

export function deserializeSequence(json: string): ActionSequence {
  return validateSequence(JSON.parse(json));
}

export function createSampleSequence(): ActionSequence {
  return validateSequence({
    version: SCHEMA_VERSION,
    name: 'Sample',
    description: 'Demo sequence',
    actions: [
      { type: 'move_mouse', x: 100, y: 100, duration: 0.3, label: 'Move to start' },
      { type: 'click', label: 'Click' },
      { type: 'type_text', text: 'Hello ForgeFlow', interval: 0.05, label: 'Type greeting' },
      { type: 'wait', seconds: 1, label: 'Pause' },
      { type: 'press_key', key: 'enter', label: 'Press Enter' },
      { type: 'hotkey', keys: ['ctrl', 's'], label: 'Save' },
      { type: 'scroll', amount: -3, label: 'Scroll down' },
      { type: 'open_application', target: 'notepad', label: 'Open Notepad' },
      { type: 'double_click', x: 200, y: 200, label: 'Double click' },
      { type: 'right_click', x: 300, y: 300, label: 'Right click' },
    ],
  });
}

File: runner/forgeflow_runner/schema.py
"""Action schema validation and normalization — mirrors web/src/lib/schema.ts."""

from __future__ import annotations

import json
from dataclasses import dataclass, field
from typing import Any, Literal, Union

SCHEMA_VERSION = "1.0"

SAFETY_LIMITS = {
    "max_mouse_duration": 30.0,
    "min_mouse_duration": 0.05,
    "max_type_interval": 2.0,
    "min_type_interval": 0.01,
    "min_wait_seconds": 0.05,
    "max_wait_seconds": 300.0,
}


class SchemaValidationError(ValueError):
    pass


def _clamp(value: float, minimum: float, maximum: float) -> float:
    return max(minimum, min(maximum, value))


@dataclass
class MoveMouseAction:
    type: Literal["move_mouse"] = "move_mouse"
    x: float = 0
    y: float = 0
    duration: float = 0.5
    label: str | None = None


@dataclass
class ClickAction:
    type: Literal["click"] = "click"
    x: float | None = None
    y: float | None = None
    button: Literal["left", "right", "middle"] = "left"
    label: str | None = None


@dataclass
class DoubleClickAction:
    type: Literal["double_click"] = "double_click"
    x: float | None = None
    y: float | None = None
    label: str | None = None


@dataclass
class RightClickAction:
    type: Literal["right_click"] = "right_click"
    x: float | None = None
    y: float | None = None
    label: str | None = None


@dataclass
class TypeTextAction:
    type: Literal["type_text"] = "type_text"
    text: str = ""
    interval: float = 0.05
    label: str | None = None


@dataclass
class PressKeyAction:
    type: Literal["press_key"] = "press_key"
    key: str = ""
    label: str | None = None


@dataclass
class HotkeyAction:
    type: Literal["hotkey"] = "hotkey"
    keys: list[str] = field(default_factory=list)
    label: str | None = None


@dataclass
class WaitAction:
    type: Literal["wait"] = "wait"
    seconds: float = 1.0
    label: str | None = None


@dataclass
class OpenApplicationAction:
    type: Literal["open_application"] = "open_application"
    target: str = ""
    label: str | None = None


@dataclass
class ScrollAction:
    type: Literal["scroll"] = "scroll"
    amount: int = 0
    x: float | None = None
    y: float | None = None
    label: str | None = None


ForgeAction = Union[
    MoveMouseAction,
    ClickAction,
    DoubleClickAction,
    RightClickAction,
    TypeTextAction,
    PressKeyAction,
    HotkeyAction,
    WaitAction,
    OpenApplicationAction,
    ScrollAction,
]


@dataclass
class ActionSequence:
    version: str
    actions: list[ForgeAction]
    name: str | None = None
    description: str | None = None


def normalize_action(action: ForgeAction) -> ForgeAction:
    if isinstance(action, MoveMouseAction):
        action.duration = _clamp(
            action.duration,
            SAFETY_LIMITS["min_mouse_duration"],
            SAFETY_LIMITS["max_mouse_duration"],
        )
    elif isinstance(action, TypeTextAction):
        action.interval = _clamp(
            action.interval,
            SAFETY_LIMITS["min_type_interval"],
            SAFETY_LIMITS["max_type_interval"],
        )
    elif isinstance(action, WaitAction):
        action.seconds = _clamp(
            action.seconds,
            SAFETY_LIMITS["min_wait_seconds"],
            SAFETY_LIMITS["max_wait_seconds"],
        )
    return action


def _parse_action(data: dict[str, Any]) -> ForgeAction:
    action_type = data.get("type")
    if not action_type:
        raise SchemaValidationError("Action type is required")

    label = data.get("label")

    if action_type == "move_mouse":
        if "x" not in data or "y" not in data:
            raise SchemaValidationError("move_mouse requires x and y")
        return normalize_action(
            MoveMouseAction(
                x=float(data["x"]),
                y=float(data["y"]),
                duration=float(data.get("duration", 0.5)),
                label=label,
            )
        )
    if action_type == "click":
        return normalize_action(
            ClickAction(
                x=data.get("x"),
                y=data.get("y"),
                button=data.get("button", "left"),
                label=label,
            )
        )
    if action_type == "double_click":
        return DoubleClickAction(x=data.get("x"), y=data.get("y"), label=label)
    if action_type == "right_click":
        return RightClickAction(x=data.get("x"), y=data.get("y"), label=label)
    if action_type == "type_text":
        if "text" not in data:
            raise SchemaValidationError("type_text requires text")
        return normalize_action(
            TypeTextAction(
                text=str(data["text"]),
                interval=float(data.get("interval", 0.05)),
                label=label,
            )
        )
    if action_type == "press_key":
        if "key" not in data:
            raise SchemaValidationError("press_key requires key")
        return PressKeyAction(key=str(data["key"]), label=label)
    if action_type == "hotkey":
        keys = data.get("keys", [])
        if not isinstance(keys, list) or len(keys) < 2:
            raise SchemaValidationError("hotkey requires at least 2 keys")
        return HotkeyAction(keys=[str(k) for k in keys], label=label)
    if action_type == "wait":
        if "seconds" not in data:
            raise SchemaValidationError("wait requires seconds")
        return normalize_action(WaitAction(seconds=float(data["seconds"]), label=label))
    if action_type == "open_application":
        if "target" not in data:
            raise SchemaValidationError("open_application requires target")
        return OpenApplicationAction(target=str(data["target"]), label=label)
    if action_type == "scroll":
        if "amount" not in data:
            raise SchemaValidationError("scroll requires amount")
        return ScrollAction(
            amount=int(data["amount"]),
            x=data.get("x"),
            y=data.get("y"),
            label=label,
        )

    raise SchemaValidationError(f"Unknown action type: {action_type}")


def validate_sequence(data: dict[str, Any]) -> ActionSequence:
    actions_raw = data.get("actions")
    if not isinstance(actions_raw, list) or len(actions_raw) == 0:
        raise SchemaValidationError("Sequence must have at least one action")

    actions = [_parse_action(a) for a in actions_raw]
    return ActionSequence(
        version=SCHEMA_VERSION,
        name=data.get("name"),
        description=data.get("description"),
        actions=actions,
    )


def serialize_sequence(sequence: ActionSequence) -> str:
    def action_to_dict(action: ForgeAction) -> dict[str, Any]:
        result: dict[str, Any] = {"type": action.type}
        if action.label:
            result["label"] = action.label
        if isinstance(action, MoveMouseAction):
            result.update({"x": action.x, "y": action.y, "duration": action.duration})
        elif isinstance(action, ClickAction):
            if action.x is not None:
                result["x"] = action.x
            if action.y is not None:
                result["y"] = action.y
            result["button"] = action.button
        elif isinstance(action, (DoubleClickAction, RightClickAction)):
            if action.x is not None:
                result["x"] = action.x
            if action.y is not None:
                result["y"] = action.y
        elif isinstance(action, TypeTextAction):
            result.update({"text": action.text, "interval": action.interval})
        elif isinstance(action, PressKeyAction):
            result["key"] = action.key
        elif isinstance(action, HotkeyAction):
            result["keys"] = action.keys
        elif isinstance(action, WaitAction):
            result["seconds"] = action.seconds
        elif isinstance(action, OpenApplicationAction):
            result["target"] = action.target
        elif isinstance(action, ScrollAction):
            result["amount"] = action.amount
            if action.x is not None:
                result["x"] = action.x
            if action.y is not None:
                result["y"] = action.y
        return result

    payload = {
        "version": SCHEMA_VERSION,
        "actions": [action_to_dict(a) for a in sequence.actions],
    }
    if sequence.name:
        payload["name"] = sequence.name
    if sequence.description:
        payload["description"] = sequence.description
    return json.dumps(payload, indent=2)


def deserialize_sequence(json_str: str) -> ActionSequence:
    return validate_sequence(json.loads(json_str))


def action_to_dict(action: ForgeAction) -> dict[str, Any]:
    """Public helper for tests and executor."""
    return json.loads(
        serialize_sequence(ActionSequence(version=SCHEMA_VERSION, actions=[action]))
    )["actions"][0]